func: teleport and collision-fast, needs wall dims

rishyfishy opened the pull request:
Added teleport function (returns whether teleport is valid, i.e. within map and no collisions)

rishyfishy committed:
func: teleport and collision-fast, needs wall dims

rishyfishy committed:
revert to original robot (for formatting)

rishyfishy committed:
feat:  functions to simplify map, performance improvements, and changed add_error

rishyfishy committed:
feat: faster check_collision_walls_fast function

rishyfishy committed:
feat: added maze_dim_x and y

rishyfishy committed:
fix:added missing check_collision_fast function

rishyfishy committed:
feat:faster collision checking and new teleport function

rishyfishy committed:
feat: simulateFast function for faster sensing

rishyfishy committed:
feat: reduce wall count by merging colinear intersecting walls

rishyfishy committed:
fix: simulate now simulates fast. Old version of simulate commented out

rishyfishy committed:
test: 8 ppi

rishyfishy committed:
feat: remove shared walls of neighbouring blocks

rishyfishy committed:
fix:plot reduced wall segments

rishyfishy committed:
feat: Using reduced wall set to calculate robot collisions

rishyfishy committed:
Merge branch 'ian612:main' into performanceImporvements

rishyfishy commented:
Minimized the number of walls in the game.
- removes shared walls of neighboring blocks (both instances)
- merges any other collinear intersecting lines to reduce number of lines

Improved performance of collision detection
- faster algorithm for calculating existence of collision (as compared to calculating the actual point) is used when coordinate is not needed

Improved performance of ultrasonic sensor
- only calculates distance of nearest collision rather than doing multiple sqrt calculations.

rishyfishy committed:
removed newtonraphson and improved commenting

rishyfishy committed:
Merge branch 'performanceImporvements' of https://github.com/rishyfishy/simmer-python into performanceImporvements

ian612 commented on devices/ultrasonic.py:
```diff
@@ -127,28 +125,70 @@ def simulate(self, value: float, environment: dict):
+        
         return utilities.add_error(output, self.error_pct, self.reading_bounds)
 
+    ### old version of simulate(). Delete if new one is stable.
```
When you're using git, it's not actually necessary to do this kind of thing because git stores all the different versions. You can do it temporarily while testing, but in general you should remove it before submitting it in a pull request.

Can you please remove this commented out function? Thanks!

ian612 commented on devices/ultrasonic.py:
```diff
@@ -107,17 +107,15 @@ def draw_measurement(self, canvas):
-
     def simulate(self, value: float, environment: dict):
-        '''
+        """
```
For consistency with other code, use single quotes for docstrings

ian612 commented on devices/ultrasonic.py:
```diff
@@ -127,28 +125,70 @@ def simulate(self, value: float, environment: dict):
-
-            for square in to_check:
-                for segment_wall in square:
-                    collision_points = utilities.collision(ray, segment_wall)
+            # if self._block_visible(BLOCK):
+            #     to_check = [BLOCK.block_square, *MAZE.wall_squares]
+            # else:
```
Does this function no longer check to see if the block is detected by the ultrasonic sensor? If so, is that functionality replaced?

rishyfishy replied:
This was a mistake on my part. I will fix it

ian612 commented on devices/ultrasonic.py:
```diff
@@ -127,28 +125,70 @@ def simulate(self, value: float, environment: dict):
-                        rays[ct][1], ray_lengths[ct] = utilities.closest(self.position_global, collision_points)
+                        rays[ct][1], ray_lengths[ct] = utilities.closestFast(
+                            self.position_global, collision_points
+                        )
```
Remove extra tabs

```suggestion
            for wall in MAZE.reduced_walls:
                collision_points = utilities.collision(ray, wall)
                if not collision_points:
                    pass
                else:
                    rays[ct][1], ray_lengths[ct] = utilities.closestFast(
                        self.position_global, collision_points
                    )
```

ian612 commented on config.py:
```diff
@@ -63,10 +63,13 @@
 
 # Graphics information
 frame_rate = 60             # Target frame rate (Hz)
-ppi = 16                    # Number of on-screen pixels per inch on display
```
Revert this to 16 as the default

ian612 commented on utilities.py:
```diff
@@ -22,9 +22,10 @@
-def add_error(value: float, pct_error: float, bounds: list = []):
-    '''
+def add_error(value: float, pct_error: float, bounds: list = [],sigma=2):
+    """
```
revert to '''

ian612 commented on utilities.py:
```diff
@@ -171,10 +174,319 @@ def closest(start: list, test_pts: list):
+                distSq_minimum = distSq
+                closest_pt = test_pt
+
+    #TODO: potentially only calculate sqrt once rather than 7 times per sensor
```
Is this completed by the `closestFast()` function?

rishyfishy replied:
It is now. Instead of storing the ray_lengths in the ultrasonic sensor object, I store ray_lengths_squared, and only squareroot the smallest one when returning the reading

rishyfishy replied:
I am a little worried that performance improvements beyond this point would sacrifice readability, and even this one I'm not if the readability - performance gain trade-off is worth it. I didn't want to refactor too many things to make it more readable because my idea of readability might be different from someone else's

ian612 commented on robot.py:
```diff
@@ -190,22 +190,36 @@ def move(self, velocity, rotation, walls):
+                
     def check_collision_walls(self, walls: list):
-        '''
+        """
```
revert to single quotes

ian612 commented on robot.py:
```diff
@@ -190,22 +190,36 @@ def move(self, velocity, rotation, walls):
         Checks for a collision between the robot's perimeter segments
         and a set of wall line segments.
-        '''
+        """
```
single quotes

ian612 commented on utilities.py:
```diff
@@ -171,10 +174,319 @@ def closest(start: list, test_pts: list):
+        return 180 - diff
+
+
+def closestFast(start: list, test_pts: list):
```
for Python variables and function names, convention is to use `snake_case` rather than `camelCase`.

ian612 reviewed (changes requested):
Thanks so much for contributing your modifications! I've marked a few things for review and left a few comments on things to fix, then I'll merge these into the repository. Overall, really nice stuff that I'm sure will make the simulator more useful for future students.

Most of it is small style things, but I had a couple questions as well about how the new code works.

The main bug I found is that I don't think the block is being detected by ultrasonic sensors anymore (at least in my test case I had the measurement rays going through it).

![image](https://github.com/ian612/simmer-python/assets/34078802/1f45e125-a801-484f-91f7-777b06749981)

rishyfishy committed:
I think I fixed it

rishyfishy committed:
renamed functions to snake case

rishyfishy committed:
single quotes for docstrings (Undoing black formatters mistakes)

rishyfishy committed:
Measure the actual distance of only the shortest ray

rishyfishy committed:
code formatting single quotes

rishyfishy commented:
Hey Ian. Fixed the formatting issues and the bugs you mentioned. I also added the change where it calculates the sqrt only for the shortest beam rather than all the beams' distances. Sorry for the delays, I didn't see some of your comments until recently

ian612 committed:
teleport function now reverts position if collision detected

ian612 commented:
Apologies for the delay in getting this integrated, but I've made a couple of tweaks and will merge this now. Thanks again for making these updates!
